Jet tree maker updates with additional TOF information saved in the trees

zlchang opened the pull request.

zlchang committed:
update jet maker with TOF information from Bassam

zlchang committed:
update jet maker with TOF information from Bassam

plexoos commented on StRoot/StJetMaker/StJetMaker2009.cxx:
```diff
@@ -91,8 +91,18 @@ int StJetMaker2009::Make()
       // Keep track of number of good vertices
       int nvertices = 0;
 
+      // Start: bassam
```
I would not add such comments in the code as I don't see much value in them. For an onlooker as myself these just add noise to the code when reviewing it

plexoos commented on StRoot/StJetMaker/StJetMaker2009.h:
```diff
@@ -79,7 +84,15 @@ struct StJetBranch {
+  Bool_t  mStoreOnlyDefaultVertex;
+  // End: bassam
+
+  ClassDef(StJetMaker2009,0); //zchang// bassam: this was changed from 0->1 after edits were done
```
This added comment does not seem to correspond to the actual change... The class version is not changed from 0 to 1

plexoos commented on StRoot/StJetMaker/StJetMaker2012.cxx:
```diff
@@ -20,12 +20,12 @@
 #include "StjBEMCMuDst.h"
 #include "StjEEMCMuDst.h"
-#include "StjFMSMuDst.h"
+//#include "StjFMSMuDst.h"
```
Instead of commenting out the lines just remove them. Hopefully, you have a good log message justifying the reason for that

plexoos commented on StRoot/StJetMaker/StJetMaker2012.cxx:
```diff
@@ -20,12 +20,12 @@
 #include "StjBEMCNull.h"
 #include "StjEEMCNull.h"
-#include "StjFMSNull.h"
+//#include "StjFMSNull.h"
```
```suggestion
```

plexoos commented on StRoot/StJetMaker/StJetMaker2012.cxx:
```diff
@@ -145,19 +156,19 @@ int StJetMaker2012::Make()
+	//if (jetbranch->anapars->useFms) {
+	  //StjFMSMuDst fms;
+	  //fmsEnergyList = fms.getEnergyList();
+	//}
```
```suggestion
```

plexoos commented on StRoot/StJetMaker/StJetSkimEventMaker.cxx:
```diff
@@ -242,6 +242,22 @@ Int_t StJetSkimEventMaker::Make()
+                mEvent->setClosestVpdWHit(muDst->btofHeader()->closestVpdWHit());
+                mEvent->setLatestVpdEHit(muDst->btofHeader()->latestVpdEHit());
+                mEvent->setLatestVpdWHit(muDst->btofHeader()->latestVpdWHit());
+                // End: bassam 
```
Check indentation. You may be using tabs instead of recommended spaces. See the STAR coding guidelines.

plexoos commented on StRoot/StJetMaker/emulator/StMuTrackEmu.h:
```diff
@@ -42,6 +42,11 @@ class StMuTrackEmu {
+    , _nSigmaTofPion(0)     // bassam
+    , _nSigmaTofKaon(0)     // bassam
+    , _nSigmaTofProton(0)   // bassam
+    , _nSigmaTofElectron(0) // bassam
```
Pointless comments again... If you want to give credit to someone you'd better mention them in the log message

plexoos commented on StRoot/StSpinPool/StJetSkimEvent/StJetSkimEvent.h:
```diff
@@ -288,6 +288,21 @@ class StJetSkimEvent : public TObject {
+                void setLatestVpdEHit(double latestVpdEHit) { mLatestVpdEHit = latestVpdEHit; }
+                void setLatestVpdWHit(double latestVpdWHit) { mLatestVpdWHit = latestVpdWHit; }
+                // End: bassam 
+
```
Check indentation...

plexoos commented on StRoot/StSpinPool/StJetSkimEvent/StJetSkimEvent.cxx:
```diff
@@ -350,6 +365,21 @@ StJetSkimEvent::StJetSkimEvent(const StJetSkimEvent &other) : TObject()
+    this->mClosestVpdWHit = other.closestVpdWHit();
+    this->mLatestVpdEHit = other.latestVpdEHit();
+    this->mLatestVpdWHit = other.latestVpdWHit();
+    // End: bassam
```
If you check the STAR guidelines using `this->` is not recommended

zlchang committed:
Two updates: a) additional TOF information to be saved in the jet tree developped by Bassam Aboona; b) removed the FMS towers in StJetMaker2012 to move ways for a future jet maker specifically for FMS jet analysis

zlchang commented:
Hi @plexoos,

I have removed the pointless comments "//bassam". Those were added by Bassam to better distinguish his edits. I will relay this message to him so that these line won't be needed. 

I removed the FMS tower in the StJetMaker2012, this is because we expect a new jet maker to be uploaded that is specifically written for FMS jet analysis by Latif Kabir soon.

ClassDef(StJetMaker2009, 0) is kept because this maker is not for ROOT I/O, previously Bassam incremented it by 1, and I changed it back to 0, that's why you saw "zchang" prefixed to original Bassam's comments. I removed the comment as well.

This pointer has been used in our jet code extensively before. It may be better to keep it as it is now to avoid modifying it everywhere in the code. Let me know if this is okay with you.

Indentation is wrong in multiple places even from the original code, but I fixed those indentations you mentioned in StJetSkimMaker and StJetSkimEvent.

zlchang commented:
Hi @plexoos,

I just committed an updates, but it seems the ROOT5 compilation check has failed. I can compile the code on RCF under SL21a.

plexoos commented:
Thanks for cleaning up the code Zilong. It looks better and I don't have other suggestions.
The issue with the ROOT5 build appears to be a glitch on the GitHub side and is not related to your changes. If you go to the "Checks" tab you should see a "Re-run jobs" button. Let's see if the jobs go through this time.

zlchang commented:
@akioogawa and @veprbl any comments for this code, and we need this for the run17 jet tree production?

veprbl commented:
Why PR that is supposed to add TOF information is removing FMS?

zlchang commented:
There will be another maker dedicated to FMS jet analysis from Latif uploaded soon.

On Aug 9, 2021, at 9:14 AM, Dmitry Kalinkin ***@***.***> wrote:

﻿

Why PR that is supposed to add TOF information is removing FMS?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/star-bnl/star-sw/pull/79#issuecomment-895212445>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGH7F3MHAL5P5KYGAO4FFZTT37IDJANCNFSM5BM25RFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

veprbl commented:
Latif can remove FMS then :)

zlchang commented:
I discussed this with Latif and we both agreed that we can remove FMS in the jet maker and there is also a bug for the EEMC tower when no vertex found and we will update it when his code is ready.

Sent from my iPhone

On Aug 9, 2021, at 9:40 AM, Dmitry Kalinkin ***@***.***> wrote:

﻿

Latif can remove FMS then :)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/star-bnl/star-sw/pull/79#issuecomment-895232059>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGH7F3MLGL4GI56VKLDXKWTT37LFTANCNFSM5BM25RFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

veprbl commented:
So is FMS being removed or not? There are still files left:
```
# find StRoot/StJetMaker -iname "*FMS*"                         
StRoot/StJetMaker/towers/StjFMSTree.h
StRoot/StJetMaker/towers/StjFMS.h
StRoot/StJetMaker/towers/StjFMSNull.h
StRoot/StJetMaker/towers/StjFMSTree.cxx
StRoot/StJetMaker/towers/StjFMSNull.cxx
StRoot/StJetMaker/towers/StjFMS.cxx
StRoot/StJetMaker/mudst/StjFMSMuDst.cxx
StRoot/StJetMaker/mudst/StjFMSMuDst.h
```
https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/StAnaPars.h#L108
https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/macros/RunJetFinder2012pro.C#L46-L53
https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/macros/RunJetFinder2012pro.C#L80-L83

zlchang commented:
These will be used in his new maker. The jet maker is just to read the inputs from MuDst files and call the jet finder. His new maker is very similar to the current StJetMaker2012.


On Aug 9, 2021, at 10:33 AM, Dmitry Kalinkin ***@***.***> wrote:

﻿

So is FMS being removed or not? There are still files left:

# find StRoot/StJetMaker -iname "*FMS*"
StRoot/StJetMaker/towers/StjFMSTree.h
StRoot/StJetMaker/towers/StjFMS.h
StRoot/StJetMaker/towers/StjFMSNull.h
StRoot/StJetMaker/towers/StjFMSTree.cxx
StRoot/StJetMaker/towers/StjFMSNull.cxx
StRoot/StJetMaker/towers/StjFMS.cxx
StRoot/StJetMaker/mudst/StjFMSMuDst.cxx
StRoot/StJetMaker/mudst/StjFMSMuDst.h


https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/StAnaPars.h#L108
https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/macros/RunJetFinder2012pro.C#L46-L53
https://github.com/star-bnl/star-sw/blob/3d486c7bfa5074a0dff650109391414fc75a9f83/StRoot/StJetMaker/macros/RunJetFinder2012pro.C#L80-L83

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/star-bnl/star-sw/pull/79#issuecomment-895275593>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGH7F3IZS74ZMBFAD2I6UCDT37RK5ANCNFSM5BM25RFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

plexoos commented:
Zilong, Dmitry makes a good point asking you to not submit changes unrelated to the topic stated in this PR. Just create a separate PR with FMS modifications, it costs nothing but makes the intention much clear.

veprbl commented on StRoot/StJetMaker/StJetMaker2009.cxx:
```diff
@@ -91,8 +91,16 @@ int StJetMaker2009::Make()
+      }   
+      else {
+          totNumVertices = tpc.numberOfVertices();
+      }   
```
May I suggest a more concise version. Also we usually store flags in StAnaPars, let's not create alternative ways to store them.
```suggestion
      int totNumVertices = tpc.numberOfVertices(); // store the total number of available vertices
      if(jetbranch->anapars->storeOnlyDefaultVertex && tpc.numberOfVertices() > 1) {
          totNumVertices = 1;
      } 
```

veprbl commented on StRoot/StJetMaker/StJetMaker2012.cxx:
```diff
@@ -108,8 +106,16 @@ int StJetMaker2012::Make()
+      }
+      else {
+          totNumVertices = tpc.numberOfVertices();
+      }
```
ditto

veprbl commented on StRoot/StJetMaker/mudst/StjTPCMuDst.cxx:
```diff
@@ -106,6 +115,12 @@ StjTrack StjTPCMuDst::createTrack(const StMuTrack* mutrack, int i, double magnet
   track.chi2prob   = mutrack->chi2prob();
   track.BField     = magneticField;
 
+  track.btofTrayId = btofTrayId;
```
```suggestion
  track.btofTrayId = -999;
  if(mutrack->index2BTofHit() >= 0) { // to make sure that there is a BTOF hit
      track.btofTrayId = mutrack->tofHit()->tray();
  }
```

veprbl commented on StRoot/StSpinPool/StJetSkimEvent/StJetSkimEvent.h:
```diff
@@ -145,357 +145,397 @@ inline void StJetSkimTrig::setTotalEnergy(int aEnergy) {mTotalEnergy = aEnergy;}
+  void setL2Result(const int* rhs);
+  void setMcEvent(const StPythiaEvent *ptr) {mMcEvent = ptr;}
+
+  void setNTzero(short n) { mNTzero = n; }
+  void setNTzeroCan(short nCan) { mNTzeroCan = nCan; }
+  void setTCanFirst(double tFirst) { mTCanFirst = tFirst; }
+  void setTCanLast(double tLast) { mTCanLast = tLast; }
```
Would be cool if someone knowledgeable would add a comment on what those values are.
I could not find them https://www.star.bnl.gov/webdata/dox/html/classStBTofHeader.html

BassamAboona replied:
Hi @zlchang and @veprbl! Since I can't make the edits to this file here are the comments that can be added for these variables:

mNTzero: Number of particles used for start-time calculations
mNTzeroCan: Number of candidate particles available for start-time calculations before outlier rejection
mTCanFirst: Earliest candidate particle in start-time calculations
mTCanLast: Latest candidate particle in start-time calculations

Please, let me know if you would like more details.

veprbl replied:
@BassamAboona Thanks!

```suggestion
  void setNTzero(short n) { mNTzero = n; } //! Set number of particles used for start-time calculations
  void setNTzeroCan(short nCan) { mNTzeroCan = nCan; } //! Set number of candidate particles available for start-time calculations before outlier rejection
  void setTCanFirst(double tFirst) { mTCanFirst = tFirst; } //! Set earliest candidate particle in start-time calculations
  void setTCanLast(double tLast) { mTCanLast = tLast; } //! Set latest candidate particle in start-time calculations
```

veprbl commented on StRoot/StSpinPool/StJetEvent/StJetEvent.h:
```diff
@@ -82,7 +82,7 @@ class StJetEvent : public TObject {
   TClonesArray* mParticles;
 
-  ClassDef(StJetEvent,6);
+  ClassDef(StJetEvent,7); //incremented to 7 for new updates with TOF information
```
This change is not needed.
```suggestion
  ClassDef(StJetEvent,6);
```

zlchang replied:
StJetEvent structure isn't changed, but StJetTrack, I kept Bassam change originally just to hope this will add an additional alert right away if you used an older version of StJetEvent code, the jet tree may not be compatible. But I don't have strong objection to this.

veprbl replied:
@zlchang I see your point. The version numbers don't hurt or help anything, so we might use them to indicate major changes. I would not keep the comment though.

veprbl commented:
> Zilong, Dmitry makes a good point asking you to not submit changes unrelated to the topic stated in this PR. Just create a separate PR with FMS modifications, it costs nothing but makes the intention much clear.

My standard is lower than that. I'm asked to review change in context of some other change that I haven't seen. If there exists a privately discussed consensus between jet experts, I don't mind this merged, just let's not pretend that I've reviewed this.

zlchang commented:
That’s easy to do and I can put these changes back in.

On Aug 9, 2021, at 10:45 AM, Dmitri Smirnov ***@***.***> wrote:

﻿

Zilong, Dmitry makes a good point asking you to not submit changes unrelated to the topic stated in this PR. Just create a separate PR with FMS modifications, it costs nothing but makes the intention much clear.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/star-bnl/star-sw/pull/79#issuecomment-895285234>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGH7F3KRW4B3WLOHADI5BE3T37SZ5ANCNFSM5BM25RFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

zlchang commented:
There is nothing wrong with this code, it works and it has already been tested.

On Aug 9, 2021, at 11:08 AM, Dmitry Kalinkin ***@***.***> wrote:

﻿

Zilong, Dmitry makes a good point asking you to not submit changes unrelated to the topic stated in this PR. Just create a separate PR with FMS modifications, it costs nothing but makes the intention much clear.

My standard is lower than that. I'm asked to review change in context of some other change that I haven't seen. If there exists a privately discussed consensus between jet experts, I don't mind this merged, just let's not pretend that I've reviewed this.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/star-bnl/star-sw/pull/79#issuecomment-895304794>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGH7F3OT6AI54VIS7QOSG5DT37VPBANCNFSM5BM25RFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

veprbl commented on StRoot/StSpinPool/StJetSkimEvent/StJetSkimEvent.h:
```diff
@@ -145,357 +145,397 @@ inline void StJetSkimTrig::setTotalEnergy(int aEnergy) {mTotalEnergy = aEnergy;}
+  short nTzeroCan() const { return mNTzeroCan; }
+  float tStart() const {return mVpdTstart;}
+  double tCanFirst() const { return mTCanFirst; }
+  double tCanLast() const { return mTCanLast; }
```
```suggestion
  short nTzero() const { return mNTzero; } //! Get number of particles used for start-time calculations
  short nTzeroCan() const { return mNTzeroCan; } //! Get number of candidate particles available for start-time calculations before outlier rejection
  double tCanFirst() const { return mTCanFirst; } //! Get earliest candidate particle in start-time calculations
  double tCanLast() const { return mTCanLast; } //! Get latest candidate particle in start-time calculations
```

The `tStart()` is a duplicate of `vpdTstart()` above

BassamAboona replied:
@veprbl Carl and I wanted to leave the vpdTstart() function alone and not change it. However, the start-time returned by the vpdTstart() function in not necessarily always the start-time from the VPD. So, we added a start-time function that has a more generic name where there is no reference to what detector (VPD or BTOF) calculated the start-time.

veprbl replied:
This may be premature to add such function before the different behaviours are actually implemented, but okay. Can we, please, add a comment to explain your reasoning in the code?

BassamAboona replied:
@veprbl Yeah! It is totally fine with me to add a comment explaining our reasoning.

veprbl replied:
```suggestion
  //! This method returns a generic start-time. Unlike
  //! StJetSkimEvent::vpdTstart it does not specify which detector
  //! (VPD or BTOF) will be used to calculate the start-time.
  float tStart() const { return vpdTstart(); }

  short nTzero() const { return mNTzero; } //! Get number of particles used for start-time calculations
  short nTzeroCan() const { return mNTzeroCan; } //! Get number of candidate particles available for start-time calculations before outlier rejection
  double tCanFirst() const { return mTCanFirst; } //! Get earliest candidate particle in start-time calculations
  double tCanLast() const { return mTCanLast; } //! Get latest candidate particle in start-time calculations
```

zlchang replied:
@veprbl @BassamAboona, what about we removed the tStart() function, instead I added a comment saying vpdTStart() isn't necessarily coming from VPD? Two reasons I think we should remove this function: a) there isn't a set function corresponding to tStart(); b) I also checked the StJetSkimEventMaker, in that function the mVpdTstart is also set by setVpdTstart(). To make code consistent overall, removing tStart() may be a better option. In the analysis code, you can always note what this time is for. By the way, can this time come from start-less tof?

veprbl replied:
> vpdTStart() isn't necessarily coming from VPD

If it's named "vpdTStart", it better come from VPD. If not, it's an unfortunate bug.

BassamAboona replied:
@veprbl and @zlchang Yeah! This "VPD" start-time can come from start-less tof. I think historically, start-less TOF start-time was not widely used by the Spin group because the algorithm for start-less TOF was not optimized for our collision energies. So, it used to be safe to assume that the start-time was primarily coming from the VPD, and thus the name "vpdTStart()" for the function. However, with the recent work that Carl and I did, start-less TOF can provide start-time a lot more often than it used. However, an event can either have a start-time calculated by the VPD or start-less TOF and not both.

I am okay with removing the tStart() function, but users need to know, and actually dig in the code and read the comments, that "vpdTStart" doesn't mean that the start-time necessarily came from the VPD.

Alternatively, the "vpdTStart" can be renamed "tStart"

veprbl replied:
@BassamAboona I'm not aware of this. Could you please point me to where btofHeader is mangled? Is there a some kind of startless tof maker that replaces the vpd time?

BassamAboona replied:
@veprbl The process of finding the start-time is mostly done in StBTofCalibMaker and the start-time is stored in StBTofHeader. I think the best thing to do is search "mTStart" in StBTofCalibMaker. You also might want to look at the routine in there named tstart_NoVpd() on line 2050. 

StRoot/StBTofCalibMaker/StBTofCalibMaker.cxx

zlchang replied:
@BassamAboona do you know if there is a function to return VPD time in the btofheader? If not, we can rename the variable, the setter function and the getter function.

BassamAboona replied:
@zlchang There is no function in StBTofHeader that specifically returns VPD start-time. StBTofHeader has a function named tStart(). This will return the event's start-time and the returned start-time could have been calculated using start-less TOF, VPD, etc. Carl and I intended to mimic this function in the jet codes because "vpdTStart()" is misleading, but we also did not want to just modify the existing "vpdTStart()". 

In other words, if an event has a start-time, then it is stored in the same spot in StBTofHeader regardless of what method was used to obtain the start-time. One event can not have two start-times that are calculated using different methods, which then makes "vpdTStart()" misleading if the start-time was actually calculated using start-less TOF.

veprbl replied:
> @BassamAboona do you know if there is a function to return VPD time in the btofheader? If not, we can rename the variable, the setter function and the getter function.

It seems like it's there isn't. It's either `StBTofCalibMaker::tstart` or `StBTofCalibMaker::tstart_NoVpd` and the later ignores VPD time information from `mTSumEast`/`mTSumWest`. I think we should just get rid of the `StJetSkimEvent::vpdTStart()`.

BassamAboona replied:
@veprbl Right! It is either start-less TOF or VPD start-time and not both.

zlchang committed:
Updates with TOF information passed into jet maker from MuDst files and saved in StJetTrack and StJetSkimEvent

zlchang committed:
changed the mVpdTstart in StJetSkimEvent to mTstart, as well as the name of its getter and setter, to avoid confusion, because the btofheader returns either the VPD start time or the start time from start-less TOF

zlchang commented:
I pushed a new set of commit, let me know if this is okay with you guys.

zlchang committed:
some cleanups

BassamAboona commented:
Hi Zilong! The edits related to TOF information look good to me. Thank you very much for making these edits!

zlchang commented:
Are we ready to move forward with this pull request?

veprbl committed:
whitespace cleanup

veprbl committed:
StJetMaker{2009,2012}: refactor to not call tpc.numberOfVertices() twice

veprbl reviewed (approved):
Diff looks fine

plexoos commented:
When merging this PR please use "Squash and merge" in order to keep the main branch clean
